mirgecom/artificial_viscosity.py

```diff
@@ -0,0 +1,340 @@
+
+.. math::
+
+    S_e = \frac{\langle u_{N_p} - u_{N_{p-1}}, u_{N_p} -
+        u_{N_{p-1}}\rangle_e}{\langle u_{N_p}, u_{N_p} \rangle_e}
+
+where:
+
+- $S_e$ is the smoothness indicator
+- $u_{N_p}$ is the solution in modal basis at the current polynomial order
+- $u_{N_{p-1}}$ is the truncated modal represention to the polynomial order $p-1$
+- The $L_2$ inner product on an element is denoted $\langle \cdot,\cdot \rangle_e$
+
+The elementwise viscosity is then calculated:
+
+.. math::
+
+    \varepsilon_e =
+        \begin{cases}
+            0, & s_e < s_0 - \kappa \\
+            \frac{1}{2}\left( 1 + \sin \frac{\pi(s_e - s_0)}{2 \kappa} \right ),
+            & s_0-\kappa \le s_e \le s_0+\kappa \\
+            1, & s_e > s_0+\kappa
+        \end{cases}
+
+where:
+
+- $\varepsilon_e$ is the element viscosity
+- $s_e = \log_{10}{S_e} \sim 1/p^4$ is the smoothness indicator
+- $s_0$ is a reference smoothness value
+- $\kappa$ controls the width of the transition between 0 to 1
+
+Smoothness Indicator Evaluation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. autofunction:: smoothness_indicator
+
+AV RHS Evaluation
+^^^^^^^^^^^^^^^^^
+
+.. autofunction:: av_laplacian_operator
+"""
+
+__copyright__ = """
+Copyright (C) 2020 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+
+from pytools import memoize_in, keyed_memoize_in
+from functools import partial
+from meshmode.dof_array import thaw, DOFArray
+
+from mirgecom.flux import num_flux_central
+from mirgecom.operators import div_operator
+
+from grudge.trace_pair import (
+    interior_trace_pairs,
+    tracepair_with_discr_tag
+)
+
+from grudge.dof_desc import (
+    DOFDesc,
+    as_dofdesc,
+    DD_VOLUME_MODAL,
+    DD_VOLUME
+)
+
+import grudge.op as op
+
+
+class _AVCVTag:
+    pass
+
+
+class _AVRTag:
+    pass
+
+
+def av_laplacian_operator(discr, boundaries, fluid_state, alpha, gas_model=None,
```

Comment:
Let me ask the elephant-in-the-room question here: Given that we have variable-coefficient diffusion, why do we want this? As @majosm can possibly attest, getting diffusion with elementwise-discontinuous coefficients right is not entirely trivial. And the already-implemented diffusion operator literally implements the same operator. It doesn't help that this operator is not well-tested.

I'd be OK with this becoming a deprecated wrapper of the diffusion operator.